doc: more complete usage explanation in readme (#121)
AmadeusITGroup/tansu: README.md

Implementation wise, it is a tiny (1300 LOC) library without any external dependencies.

## Installation

You can add Tansu to your project by installing the `@amadeus-it-group/tansu` package using your favorite package manager, ex.:

* `yarn add @amadeus-it-group/tansu`
* `npm install @amadeus-it-group/tansu`

## Usage

Check out the [Tansu API documentation](https://amadeusitgroup.github.io/tansu/).

> The APIs to manage your reactive state can be categorized into three distinct groups:

Reviewer: ```suggestion
The functional part of the API to manage your reactive state can be categorized into three distinct groups:
```
(as you do not mention the class-based part here)

#### Single store

```typescript
import {writable, derived} from "@amadeus-it-group/tansu";

const value$ = writable(1);
const double$ = derived(value$, (value) => value * 2);

double$.subscribe((double) => console.log('Double value', double));
value$.set(2);
```

output:

```text
Double value 2
Double value 4
```

#### Multiple stores

```typescript
import {writable, derived} from "@amadeus-it-group/tansu";

const a$ = writable(1);
const b$ = writable(1);
const sum$ = derived([a$, b$], ([a, b]) => a + b);

sum$.subscribe((sum) => console.log('Sum', sum));
a$.set(2);
```

output:

```text
Sum 2
Sum 3
```

#### Asynchronous set

A `derived` can directly manipulate its value using the set method instead of relying on the returned value of the provided function.
This flexibility allows you to manage asynchronous operations or apply filtering logic before updating the observable's value.

```typescript
import {writable, derived} from "@amadeus-it-group/tansu";

const a$ = writable(0);
const asynchronousDouble$ = derived(a$, (a, set) => {
  setTimeout(() => set(a * 2));

> });

Reviewer: When using the asynchronous version of `derived`, the third argument is mandatory.

const evenOnly$ = derived(a$, (a, set) => {

>   if (a !== undefined && a % 2 === 0) {

Reviewer: Why do you compare `a` with undefined?

      set(a);
  }

> }, <number>undefined);

Reviewer: I don't think it is a good idea to show an example that lies about the type of `evenOnly$`. If it can be initially `undefined`, this should be visible in the type.

```suggestion
}, undefined as number | undefined);
```

asynchronousDouble$.subscribe((double) => console.log('Double (asynchronous)', double));
evenOnly$.subscribe((value) => console.log('Even', value));

a$.set(1);
a$.set(2);
```

output:

```text
Even 0
Even 2
Double 0
Double 2

> Double 4

Reviewer: The output of your example is wrong. When subscribing, the listener is called synchronously (even if you do some asynchronous stuff in the derived).

Author: I ran the code locally (with the new changed code as said before), so that I will output the right one ;)

Thanks @divdavem !

```

### computed

[api documentation](https://amadeusitgroup.github.io/tansu/functions/computed.html)

A `computed` store is another variant of a derived store, with the following characteristics:

  - **Implicit Dependencies:** Unlike in a derived store, there is no requirement to explicitly declare dependencies.

>   - **Dynamic Dependency Listening:** Dependencies are determined based on their usage. This implies that a dependency store not actively used is not automatically "listened" to, optimizing resource utilization.

Reviewer: ```suggestion
  - **Dynamic Dependency Listening:** Dependencies are determined based on their usage. This implies that a dependency not actively used is not automatically "listened" to, optimizing resource utilization.
```

#### Switch map

> This capability to subscribe/unsubscribe to the dependency stores allows to create switch maps in a natural way.

Reviewer: ```suggestion
This capability to subscribe/unsubscribe to the dependencies allows to create switch maps in a natural way.
```

    return b$();
  }
});

computedValue$.subscribe((value) => console.log('Computed value:', value));
a$.set(2);
switchToA$.set(false);
a$.set(3);
a$.set(4);
switchToA$.set(true);

```

output:

```text
Return a$
Computed value: 1
Return a$
Computed value: 2
Return b$
Computed value: 0
Return a$
Computed value: 4
```

When `switchToA$.set(false)` is called, the subscription to `a$` is canceled, which means that subsequent changes to `a$` will no longer trigger the calculation., which is only performed again when switchToA$ is set back to true.

### readable

[api documentation](https://amadeusitgroup.github.io/tansu/functions/readable.html)

Similar to Svelte stores, this function generates a store where the value cannot be externally modified.

```typescript
import {readable} from '@amadeus-it-group/tansu';

const time = readable(new Date(), (set) => {
  const interval = setInterval(() => {
    set(new Date());
  }, 1000);

  return () => clearInterval(interval);
});
```

### derived vs computed

While derived and computed may appear similar, they exhibit distinct characteristics that can significantly impact effectiveness based on use-cases:

- **Declaration of Dependencies:**

>   - `computed`: No explicit declaration of store dependencies is required, providing more flexibility in code composition.

Reviewer: ```suggestion
  - `computed`: No explicit declaration of dependencies is required, providing more flexibility in code composition.
```

  - `derived`: Requires explicit declaration of dependencies.

- **Performance:**
  - `computed`: Better performance by re-running the function only based on changes in the stores involved in the last run.
  - `derived`: Re-run the function each time a dependent store changes.

- **Asynchronous State:**
  - `computed`: Unable to manage asynchronous state.
  - `derived`: Can handle asynchronous state with the `set` method.

- **Skipping Value Emission:**
  - `computed`: Does not provide a mechanism to skip emitting values.
  - `derived`: Allows skipping the emission of values by choosing not to call the provided `set` method.

- **Setup and Teardown:**
  - `computed`: Lacks explicit setup and teardown methods.
  - `derived`: Supports setup and teardown methods, allowing actions such as adding or removing DOM listeners.

> 

Reviewer: Note that a specific case related to setup and teardown can be mentioned: if a computed or derived store has a listener that unsubscribes and re-subscribes while the dependencies did not change, the computed is not called again (the previous value is considered up-to-date because the dependencies did not change) whereas the derived is called again (because it may, for example, add or remove DOM listeners).

While `computed` feels more intuitive in many use-cases, `derived` excels in scenarios where `computed` falls short, particularly in managing asynchronous state and providing more granular control over value emissions.

Carefully choosing between them based on specific requirements enhances the effectiveness of state management in your application.

### Getting the value

There are three ways for getting the value of a store:

```typescript
import {writable, get} from "@amadeus-it-group/tansu";

const count$ = writable(1);
const unsubscribe = count$.subscribe((count) => {
  // Will be called with the updated value, synchronously first, then each time count$ changes

>   // must be unsubscribe to end the future calls.

Reviewer: ```suggestion
  // unsubscribe must be called to prevent future calls.
```

  console.log(count);
});

// A store is also a function that you can call to get the instant value.
console.log(count$());

// Equivalent to
console.log(get(count$));
```

> [!NOTE]
> Getting the instant value implies the subscription and unsubription on the store:
>   - It can be important to know in case of setup/teardown functions.
>   - In the same scope, prefer to store the value once in a local variable instead of calling `store$()` several times.
>

> > `get` can work with Class variant of a store,

Reviewer: ```suggestion
> `get` can work with the class variant of a store.
>
```

Author: I'm going to remove it completely I think. The point is that, as @ExFlo mentioned it to me, the class usage is not mentioned in this part. It's a bit confusing. It's probably to be described in another section dedicated to the Tansu classes.

> When called inside a reactive context (i.e. inside a computed), getting the value serves to know and "listen" the dependent stores.

### batch

[api documentation](https://amadeusitgroup.github.io/tansu/functions/batch.html)

> Contrary to other libraries like Angular with signals or Svelte with runes, where the callback of a subscription is executed asynchronously (with `useEffect`), we have maintained the constraint of synchronicity between the store changes and their subscriptions in Tansu.

Reviewer: `useEffect` is neither in Angular nor in Svelte but in React (and it is not linked to a signal).
You probably meant `effect` in Angular or `$effect` in Svelte.

Author: You're right.

While it is acceptable for these frameworks to defer these calls since their goals are well-known in advance (to optimize their final rendering), this is not the case for Tansu, where the goal is to be adaptable to any situation.

> The problem with synchronous subscription is that it can create "glitches". Subscribers and computed store callbacks that are run too many times can create incorrect intermediate values.

Reviewer: ```suggestion
The problem with synchronous subscriptions is that it can create "glitches". Subscribers and computed store callbacks that are run too many times can create incorrect intermediate values.
```

```typescript
import {readable} from '@amadeus-it-group/tansu';

const time = readable(new Date(), (set) => {
  const interval = setInterval(() => {
    set(new Date());
  }, 1000);

  return () => clearInterval(interval);
});
```

### derived vs computed

While derived and computed may appear similar, they exhibit distinct characteristics that can significantly impact effectiveness based on use-cases:

- **Declaration of Dependencies:**
  - `computed`: No explicit declaration of dependencies is required, providing more flexibility in code composition.
  - `derived`: Requires explicit declaration of dependencies.

- **Performance:**
  - `computed`: Better performance by re-running the function only based on changes in the stores involved in the last run.
  - `derived`: Re-run the function each time a dependent store changes.

- **Asynchronous State:**
  - `computed`: Unable to manage asynchronous state.
  - `derived`: Can handle asynchronous state with the `set` method.

- **Skipping Value Emission:**
  - `computed`: Does not provide a mechanism to skip emitting values.
  - `derived`: Allows skipping the emission of values by choosing not to call the provided `set` method.

> - **Setup and Teardown:******

Reviewer: ```suggestion
- **Setup and Teardown:**
```

  - `computed`: Lacks explicit setup and teardown methods.
  - `derived`: Supports setup and teardown methods, allowing actions such as adding or removing DOM listeners.

>   - When the last listener unsubscribes and then subscribes again, the derived function is rerun due to its setup-teardown functionality. In contrast, a computed provides the last value without recomputing if it hasn't changed in the meantime.

Reviewer: ```suggestion
  - When the last listener unsubscribes and then subscribes again, the derived function is rerun due to its setup-teardown functionality. In contrast, a computed provides the last value without recomputing if dependencies haven't changed in the meantime.
```

The fullName store successively went through different states, including an inconsistent one, as `Sherlock Lupin` does not exist! Even if it can be seen as just an intermediate state, it is **fundamental** for a state management to only manage consistent data in order to prevent issues and optimize the code.

> In Tansu, the `batch`is available to defer **synchronously** the subscribers calls, and de facto the dependent `derived` or `computed` calculation to solve all kind of multiple dependencies issues.

Reviewer: ```suggestion
In Tansu, the `batch` is available to defer **synchronously** the subscribers calls, and de facto the dependent `derived` or `computed` calculation to solve all kind of multiple dependencies issues.
```

console.log('Process end');
```

output:

```text
Arsène Lupin
Sherlock Holmes
Process end
```
> [!NOTE]

> > - Managing stores in a batch context can be quite challenging (getting store values inside a batch, subscribing in a batch, avoiding rerunning calculations outside the batch if the value has been updated inside, ...). All potential situations have been specified in the specs.

Author: Yes, right, challenging is probably misleading. The idea that I wanted to highlight is more : 1) we are aware that depending on the context, for example when the subscribers are paused, getting the right value can be tricky, 2) Don't worry, we managed all these situations 😃 . But ok, let's just explain how it works.

> - `batch` can be called inside `batch`. The subscriber calls are performed at the end of the first batch, synchronously.

### asReadable

[api documentation](https://amadeusitgroup.github.io/tansu/functions/asReadable.html)

> `asReadable` transforms a writable store into a readable one. A readable store is a store without the set and update method.

Reviewer: Note that it is not mandatory for the store passed to `asReadable` to actually be writable.

This is useful and widely used to compose a custom store:

  - The first parameter is the writable store,
  - The second parameter is an object to extend the readable store returned.

```typescript

> import {writable, asReadable} from "@angular/core";

Reviewer: ```suggestion
import {writable, asReadable} from "@amadeus-it-group/tansu";
```

function createCounter(initialValue: number) {
  const store$ = writable(initialValue);

  return asReadable(store$, {
    increment: () => store$.update((value) => value + 1),
    decrement: () => store$.update((value) => value - 1),

>     reset: () => store$.set(initialValue);

Reviewer: ```suggestion
    reset: () => store$.set(initialValue)
```

  });
}

const counter$ = createCounter(0);

> counter$.subscribe((value) => console.log('Value: ', value);

Reviewer: ```suggestion
counter$.subscribe((value) => console.log('Value: ', value));
```

counter$.increment();
counter$.reset();
counter$.set(2); // Error, set does not exist
```
output:
```text
Value: 0
Value: 1
Value: 0
(Error thrown !)
```

### asWritable

[api documentation](https://amadeusitgroup.github.io/tansu/functions/asWritable.html)

`asWritable` is the opposite of an `asReadable`: it transforms a readable store (without a `set` method) into a writable one.

It's useful when you want to connect your computed store to the original one, or implement a custom `set` method

```typescript

> import {writable, asWritable} from "@angular/core";

Reviewer: ```suggestion
import {writable, asWritable} from "@amadeus-it-group/tansu";
```

// JSON transform errors are not managed here, to keep the example clear.
function createLocaleStorage(key: string) {

  function updateFromStorage(e) {
    if (e.key === key) {
      store$.set(JSON.parse(e.newValue));
    }
  }

  const initialValue = JSON.parse(localStorage.getItem(key) ?? '{}');
  const store$ = writable(initialValue, () => {
    window.addEventListener('storage', updateFromStorage);
    return () => window.removeEventListener('storage', updateFromStorage);
  });

>   const api = asWritable(store$, {

Reviewer: ```suggestion
  return asWritable(store$, {
```

    set(json: any) {
      localStorage.setItem(key, JSON.stringify(json));
      store$.set(json);
    },
    reset: () => api.set(initialValue),
  });

> }

Reviewer: There is a missing `return api;`

const storageValue$ = createLocaleStorage('myKey');

> storageValue$.subscribe((json) => console.log('Value: ', json);

Reviewer: ```suggestion
storageValue$.subscribe((json) => console.log('Value: ', json));
```

storageValue$.set({a: 1});
storageValue$.reset();

```
output (localStorage is updated accordingly):
```text
(javascript objects)
Value: {}
Value: {a: 1}
Value: {}
```

> ## Integration in other Frameworks

Author: Yes right ! I'll change it :)

> - `batch` can be called inside `batch`. The subscriber calls are performed at the end of the first batch, synchronously.
> - Retrieving the immediate value of a store (using any of the three methods mentioned earlier: calling `subscribe` with a subscriber that will be called with the value synchronously, using `get` or calling the store as a function) always provides the value based on the up-to-date values of all dependent stores (even if this requires re-computations of a computed or a derived inside the batch)
> - all calls to subscribers (excluding the first synchronous call during the subscribe process) are deferred until the end of the batch
> - if a subscriber has already been notified of a new value inside the batch (for example, when it is a new subscriber registered within the batch), it won't be notified again at the end of the batch if the value remains unchanged. Subscribers are invoked only when the store's value has changed since their last call.

### asReadable

[api documentation](https://amadeusitgroup.github.io/tansu/functions/asReadable.html)

> `asReadable` transforms a store into a readable one. A readable store is a store without the set and update method.

Reviewer: Actually, this is not rigorously exact. Technically, a writable store (i.e. implementing the `WritableSignal` interface) is also  a readable store (i.e. it implements the `ReadableSignal` interface).

This is useful and widely used to compose a custom store:

  - The first parameter is the writable store,
  - The second parameter is an object to extend the readable store returned.

```typescript
import {writable, asReadable} from "@amadeus-it-group/tansu";

function createCounter(initialValue: number) {
  const store$ = writable(initialValue);

  return asReadable(store$, {
    increment: () => store$.update((value) => value + 1),
    decrement: () => store$.update((value) => value - 1),
    reset: () => store$.set(initialValue)
  });
}

const counter$ = createCounter(0);

counter$.subscribe((value) => console.log('Value: ', value));

counter$.increment();
counter$.reset();
counter$.set(2); // Error, set does not exist
```
output:
```text
Value: 0
Value: 1
Value: 0
(Error thrown !)
```

### asWritable

[api documentation](https://amadeusitgroup.github.io/tansu/functions/asWritable.html)

> `asWritable` is the opposite of an `asReadable`: it transforms a readable store (without a `set` method) into a writable one.

Reviewer: Actually, this is not rigorously exact. `asWritable` does mostly the same thing as `asReadable` (and is implemented as a call to `asReadable`).
While `asReadable` returns a store that only implements the `ReadableSignal` interface (plus the extra methods passed as a parameter), `asWritable` returns a store that only implements the `WritableSignal` interface (which is an extension of the `ReadableSignal` interface), also plus the extra methods passed as a parameter.

> While being fairly minimal, this example demonstrates most of the Tansu APIs.

Reviewer: ```suggestion
While being fairly minimal, this example demonstrates most of the Tansu APIs with Angular.
```
